File: ex4_NN_MNIST/ex4_files/matlab_files/FF.m
classdef FF < handle
    
    properties
        weights
    end
    
    methods
        function obj = FF(layerDims)
            %   layerDims is an array of [n_0,..,n_M]. Note that n_0 is the
            %   input dimension and n_M is the output dimension
            n_weights = length(layerDims)-1;
            obj.weights = cell(n_weights,1);
            for i=1:n_weights
                obj.weights{i} = 0.1*randn(layerDims(i+1),layerDims(i));
            end
        end
        
        function yhat = predict(obj,x)
            a = x;
            for i=1:length(obj.weights)
                a = obj.activation(obj.weights{i}*a);
            end
            yhat = a;
        end
        
        function [steps,test_acc] = sgd(obj, X, y, epochs, eta, mb_size, Xtest, ytest)
            N = size(X,2);
            n_mbs = ceil(N/mb_size);
            acc = obj.eval_test(Xtest,ytest);
            
            updates = 0;
            steps = [updates];
            test_acc = [acc];
            
            fprintf('Starting training, test accuracy: %f\n',acc);
            for i=1:epochs
                perm = randperm(N);
                for j=0:(n_mbs-1)
                    idxs = ((mb_size*j)+1):min(((mb_size)*(j+1)),N);
                    X_mb = X(:,perm(idxs));
                    y_mb = y(:,perm(idxs));
                    grads = obj.backprop(X_mb,y_mb);
                    
                    for k=1:length(obj.weights)
                        obj.weights{k} = obj.weights{k} - (eta/mb_size)*grads{k};
                    end
                    updates = updates+1;
                    
                    if mod(updates,50)==0
                        steps = [steps updates];
                        acc = obj.eval_test(Xtest,ytest);
                        test_acc = [test_acc acc];
                    end
                    
                end
                acc = obj.eval_test(Xtest,ytest);
                fprintf('Done epoch %d, test accuracy: %f\n',i,acc);
            end
            steps = steps/n_mbs;
        end
        
        
        function grads = backprop(obj,X,y)
            
            % X is a matrix of size input_dim*mb_size
            % y is a matrix of size output_dim*mb_size
            % you should return a cell 'grads' of dims (length(weights),1) such
            % that grads{i} is a matrix containing the gradients of the
            % loss with respect to weights{i}.
            
            
            % ForwardPass
            
            % YOUR CODE HERE
            
            % BackwardPass:
            
            % YOUR CODE HERE
            
            % gradients:
            
            % YOUR CODE HERE
            
        end
        
        function acc = eval_test(obj,Xtest,ytest)
            ypred = obj.predict(Xtest);
            ypred = ypred==max(ypred);
            acc = mean(all((ytest==ypred)));
        end
        
    end
    
    methods(Static)
        function z=activation(x)
            z = tanh(x);
        end
        
        function z=activation_deriv(x)
            z = 1-tanh(x).^2;
        end
        
        function z=loss_deriv(output, target)
            % Derivative of loss function with respect to the activations
            % in the output layer.
            % we use quadratic loss, where L=0.5*||output-target||^2
            
            % YOUR CODE HERE
            
        end
        
    end
end

File: ex4_NN_MNIST/ff.py
import numpy as np

class FF(object):
	"""A simple FeedForward neural network"""
	def __init__(self, layerDims):
		super(FF, self).__init__()
		n_weights = len(layerDims)-1
		self.weights = []
		for i in range(n_weights):
			self.weights.append(0.1*np.random.randn(layerDims[i+1], layerDims[i]))



	def sgd(self, X, y, epochs, eta, mb_size, Xtest, ytest):
		N = X.shape[1]
		n_mbs = int(np.ceil(N/mb_size))
		acc = self.eval_test(Xtest, ytest)

		updates = 0
		steps = [updates]
		test_acc = [acc]
		print("Starting training, test accuracy: {0}".format(acc))

		for i in range(epochs):
			perm = np.random.permutation(N);
			for j in range(n_mbs):
				X_mb = X[:,perm[j*mb_size:(j+1)*mb_size]]
				y_mb = y[:,perm[j*mb_size:(j+1)*mb_size]]

				grads = self.backprop(X_mb, y_mb)

				for k,grad in enumerate(grads):
					self.weights[k] = self.weights[k] - (eta/mb_size)*grad

				updates = updates + 1
				if updates%50 == 0:
					steps.append(updates)
					test_acc.append(self.eval_test(Xtest, ytest))

			acc = self.eval_test(Xtest, ytest)
			print("Done epoch {0}, test accuracy: {1}".format(i+1, acc))

		steps = np.asarray(steps)
		steps = steps/n_mbs

		return steps, test_acc

	def backprop(self,X,y):
		# X is a matrix of size input_dim*mb_size
		# y is a matrix of size output_dim*mb_size
		# you should return a list 'grads' of length(weights) such
		# that grads[i] is a matrix containing the gradients of the
		# loss with respect to weights[i].

		# ForwardPass

		# YOUR CODE HERE
		neurons = [X]  # neurons[i] is a matrix: layer[i]_size X MB_size (S)
		inputs = []  # inputs[i] is a matrix: layer[i]_size X MB_size (H)
		g = self.activation  # activation function: from H[i] to S[i]
		dg = self.activation_deriv  # derivative of activation function
		for i in range(len(self.weights)):  # weights idx: (as showed in class)-1
			h = np.dot(self.weights[i], neurons[i])
			inputs.append(h)
			neurons.append(g(h))

		# BackwardPass
		
		# YOUR CODE HERE
		deltas = []
		delta_m = self.loss_deriv(neurons[len(neurons)-1],y) * dg(inputs[len(inputs)-1])
		deltas.append(delta_m)
		for i in range(len(self.weights)-1):  # i from
			delta = np.dot(self.weights[-(i+1)].T,deltas[i]) * dg(inputs[-(i+2)])
			deltas.append(delta)
		deltas = deltas[::-1]

		# Gradients

		# YOUR CODE HERE
		grads = []
		for i in range(len(self.weights)):
			grads.append(np.dot(deltas[i],neurons[i].T))

		return grads



	def predict(self,x):
		a = x
		for w in self.weights:
			a = FF.activation(self, np.dot(w,a))

		return a

	def eval_test(self,Xtest, ytest):
		ypred = self.predict(Xtest)
		ypred = ypred==np.max(ypred,axis=0)
		
		return np.mean(np.all(ypred==ytest,axis=0))


	def activation(self, x):
		return np.tanh(x)

	def activation_deriv(self, x):
		return 1-(np.tanh(x)**2)

	def loss_deriv(self, output, target):
		# Derivative of loss function with respect to the activations
		# in the output layer.
		# we use quadratic loss, where L=0.5*||output-target||^2
		
		# YOUR CODE HERE
		return output-target


